loading Dars from ZipInputStreams

gaboraranyossy-da opened the pull request:
Changes DAR loading to accept an `InputStream` instead of `ZipFile`. This is to avoid having to create temporary files when uploading DARs via the ledger api, or when loading them from a Jar file as a resource.

Note, that this change will make all the Zip entries from a DAR file loaded temporarily in memory, which should be acceptable.

fixes #547 

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

gaboraranyossy-da commented on daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala:
```diff
@@ -21,98 +22,121 @@ class DarReader[A](
-  private val manifestEntry = new ZipEntry("META-INF/MANIFEST.MF")
-
-  def readArchive(darFile: ZipFile): Try[Dar[A]] =
+  def readArchive(name: String, darStream: ZipInputStream): Try[Dar[A]] = {
```
I'd like to remove the `name` from here as it's only used for texts of errors. I think this module should not be aware of any dar files and if we need to log/return an error containing the dar file name, it should be done in the layer where the file is converted to an `InputStream`. What do you think?

gaboraranyossy-da replied:
We agreed with @gerolf-da to leave this for now as it is.

leo-da commented on ledger/api-server-damlonx/reference-v2/src/main/scala/com/daml/ledger/api/server/damlonx/reference/v2/ReferenceServer.scala:
```diff
@@ -46,8 +48,9 @@ object ReferenceServer extends App {
     val archivesTry = for {
-      zipFile <- Try(new ZipFile(file))
-      dar <- DarReader { case (_, x) => Try(Archive.parseFrom(x)) }.readArchive(zipFile)
+      zipInputStream <- Try(new ZipInputStream(new FileInputStream(file)))
```
Again this should be just the call to `readArchive(darFile: ZipFile): Try[Dar[A]]` which should also take care of the closing the input stream, which are not doing here.

gaboraranyossy-da commented on daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/UniversalArchiveReader.scala:
```diff
@@ -17,32 +16,41 @@ import scala.util.{Failure, Success, Try}
+  def readFile(file: File): Try[Dar[A]] =
+    for {
+      fileType <- supportedFileType(file)
+      inputStream <- fileToInputStream(file)
```
The only case I can think of which is not handled is when the InputStream is instantiated, the rest should be managed by the numerous decorating layers. I put something to handle issues at instantiation.

gaboraranyossy-da replied:
You're right, I thought the parseDar/Dalf will eventually close the stream. Also, when the stream blows up at instantiation there is nothing to close.

gaboraranyossy-da commented:
@leo-da I believe I addressed your concerns. I named the file loading method `readArchiveFromFile` to avoid method overloading. Also, it must accept `File` instead of `ZipFile` as latter cannot be transformed to a `ZipinputStream` as it seems. Please take another look.

stefanobaghino-da commented on daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala:
```diff
@@ -21,101 +23,127 @@ class DarReader[A](
+  case class InvalidZipEntry(name: String, entries: ZipEntries)
       extends RuntimeException(
-        s"Invalid zip entry: ${zipEntry.getName: String}, DAR: ${darInfo(darFile): String}")
+        s"Invalid zip entryName: ${name: String}, DAR: ${darInfo(entries): String}")
```
What is the purpose of the type hints here?

gaboraranyossy-da replied:
what type hints to you mean?

stefanobaghino-da replied:
`name: String` and `darInfo(entries): String`

gaboraranyossy-da replied:
This diagnostic failure has been already there before. I assume it helps debugging corrupted DAR files.

stefanobaghino-da replied:
I was referring to the explicit `: String` after each value in the string interpolation.

gaboraranyossy-da replied:
Oh, sorry. Good point, I don't know actually. @leo-da why do we need them? they indeed seem to be unnecessary.

S11001001 replied:
It's a safety measure; for example, suppose you change `darInfo`'s signature to return something else alongside the `String` or add some structure like `Option[String]`, or, I don't know, add a case class that wraps it? `toString` gets applied to everything silently and automatically, so your string interpolation just silently becomes wrong. The ascription assures that no type change has happened elsewhere that could have such an impact.

I'd like to write a replacement macro for `s` or a wart that only permits members of a whitelist of types to be interpolated (`String`, maybe some other safe candidates like `Int`). I'd also like a wart that forbids `toString`, but let's leave that aside for now. In the meantime, `: String` ascription is a good stopgap measure.

gaboraranyossy-da replied:
Got it. Thanks @stephencompall-DA!

leo-da commented on daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala:
```diff
@@ -21,101 +23,127 @@ class DarReader[A](
 
-  private val manifestEntry = new ZipEntry("META-INF/MANIFEST.MF")
+  def readArchiveFromFile(darFile: File) =
+    readArchive(darFile.getName, new ZipInputStream(new FileInputStream(darFile)))
```
@gaboraranyossy-da we are leaking the `ZipInputStream` create here. It should be closed.

gaboraranyossy-da replied:
good spot! it's addressed now.

leo-da commented on daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DarReader.scala:
```diff
@@ -21,101 +23,127 @@ class DarReader[A](
+    private def isPrimDalf(s: String): Boolean = s.toLowerCase.contains("-prim") && isDalf(s)
+  }
+
+  private[archive] def close(is: InputStream): Try[Unit] = Try(is.close())
```
this `close` is not used anywhere. Let's delete it or change it so it closes `Try[(Long, InputStream)]`

gaboraranyossy-da replied:
it's actually used in `parseDalfNamesFromManifest` and `parseOne`. Nevertheless I find it to be an overkill so I will just inline them.
